### ISSUE TITLE
problems_dens_sample_DM_sub2.txt.den

### ISSUE THREAD

**Issue by weiguangcui** [regonzar/paravt#4]
Hi @regonzar 

I found another problem with paravt. 
The dens changes when I am using different CPUs to do the calculation.
Please see the attached plots. Please note that, I omit out layer of these data. Total data points are within radius<7000 kpc/h, only data within radius < 5000 kpc/h is used for these plots. The center is the first position of the data points

Just in case you want to repeat my problem, I also attached data points with densities I got.
Thank you for your helps.

Best,
Weiguang

![problem_dens_cpu2](https://cloud.githubusercontent.com/assets/26856072/25858422/fe310126-34db-11e7-9a40-41e13a6ec110.png)
![problem_dens_cpu8](https://cloud.githubusercontent.com/assets/26856072/25858428/021d58d4-34dc-11e7-983a-01d6b73d3afb.png)
![problem_hist_cpu](https://cloud.githubusercontent.com/assets/26856072/25858434/04ddb8fc-34dc-11e7-8e8d-ccf6a7f5f198.png)

[data.zip](https://github.com/regonzar/paravt/files/987119/data.zip)

**Comment by regonzar** [comment]
This produced by a low BORDERFACTOR parameter, default is 1.5, when code split the volume for different tasks, the boundary buffer zone is too thin to recover all neighbors for particles in the boundary of each subvolume. 
For more non-homogeneous particle distributions, you should use a larger borderfactor value,, you could try a value between 4-10, recompile the code and repeat the plots.

**Comment by weiguangcui** [comment]
I have increased the BORDERFACTOR to 10. But the problem seems still there.

![problem_dens_cpu2_b10](https://cloud.githubusercontent.com/assets/26856072/25888467/98190538-3566-11e7-9249-7c55ccabe8a4.png)
![problem_dens_cpu8_b10](https://cloud.githubusercontent.com/assets/26856072/25888472/9cf1df62-3566-11e7-991e-c917201c054b.png)

I further checked the most dense particles:

`
In [255]: np.where(np.log10(pdens[ids2])>-4)[0]
Out[255]: array([  0,   1,   4, 113, 322, 792, 793, 815, 819, 820, 821, 822])

In [256]: np.where(np.log10(p2dens[ids2])>-4)[0]
Out[256]: array([  0,   4, 792, 793, 815, 819, 820, 821, 822])
`

Some of them give similar values, but the others not...

`In [264]: np.log10(p2dens[ids2])[array([  0,   4, 792, 793, 815, 819, 820, 821, 822])]
Out[264]: 
array([-2.89346915, -2.89346915, -3.9314431 , -3.96657624, -3.52201102,
       -3.57938423, -3.50501103, -3.52331326, -2.97839728])

In [265]: np.log10(pdens[ids2])[array([  0,   4, 792, 793, 815, 819, 820, 821, 822])]
Out[265]: 
array([-2.97839728, -2.97839728, -3.9314431 , -3.96657624, -3.52201102,
       -3.57938423, -3.50501103, -3.52331326, -2.97839728])
`
Please let me know if you need me to give a check on other things.
Thank you so much.

**Comment by regonzar** [comment]
Hi,

Did you make a clean compilation "make clean" after changing borderfactor parameter? Did increasing borderfactor to 10 improve the density computation convergence? can you show the density distribution plots for this case? If convergence is improved, probably you need to increase the value a little more.

In general if you are using non-periodic conditions, some particles in the border of your total volume would  have a different vt density for different number of processors since a border particle may have artificial neighbors very far, however this only happens to some particles at borders.

Another point is that when you have some co-linear or particles in the same plane, VT to avoid infinite vertices, add some random values to positions of that particles near the float precision to fix that, in that sense, two runs with the same input data and same number of cpu, may be not totally identical, they can differ by little numerical noise.

In your case it look like the most massive particles differ by ~3%, it is more than expected due numerical noise. 

If this is due numerical noise you could try QHULLOPTIONS "Qbb" and recompile, or even recompile with double presicion. However, i think this is due borderfactor, please check that.

**Comment by weiguangcui** [comment]
> Did you make a clean compilation "make clean" after changing borderfactor parameter? Did increasing borderfactor to 10 improve the density computation convergence? can you show the density distribution plots for this case? If convergence is improved, probably you need to increase the value a little more.

Yes, I used make clean before remake the code. The result showing in upper figure is with borderfactor = 10. I have tried 15, but the result is the same.

> In general if you are using non-periodic conditions, some particles in the border of your total volume would have a different vt density for different number of processors since a border particle may have artificial neighbors very far, however this only happens to some particles at borders.

Yes, I agree with that. And that is also why I am cut out 2 Mpc/h of thickness at outer layer before I made these plots.

> Another point is that when you have some co-linear or particles in the same plane, VT to avoid infinite vertices, add some random values to positions of that particles near the float precision to fix that, in that sense, two runs with the same input data and same number of cpu, may be not totally identical, they can differ by little numerical noise.
> In your case it look like the most massive particles differ by ~3%, it is more than expected due numerical noise.
>If this is due numerical noise you could try QHULLOPTIONS "Qbb" and recompile, or even recompile with double presicion. However, i think this is due borderfactor, please check that.

I used "Qbb" with double precision (#define REALfloat 0 in usr_r.h)

Just in case of I did some stupid unexpected changes in the code. Could you download and test my uploaded data (data.zip) in the first comment box?

Thank you so much for your helps

**Comment by regonzar** [comment]
Hi,
i've checked your data and run with borderfactor 10 and got this plot where x axis is log10(density) for 2(black) and 8(red) processors, and there is very good convergence except some border particles. However my results are totally different than yours.... can youshow plots number vs density for 2 and 8 processors using borderfactor=10.

I think there is something in your modified code. I made several test on large data-set from2 to 256 task and there is very good converge in the code. 

![fig](https://cloud.githubusercontent.com/assets/6602575/25915388/52dfef44-358f-11e7-9967-92bf4cace746.jpg)

**Comment by weiguangcui** [comment]
I redownload your code and did not change anything besides the attached config.h file.
I still can not get your results, not even closer! Your density is peaked at ~-5, but mine is much smaller ~ -8.5. However, their distributions are similar now. 

But this is not main problem, the main problem is that the particle at the center (first particle position) has always much higher density (0.001278 for CPU 2; 0.001051 for CPU 8) which should not be correct.
The nearest particle of it is 613.12527 away. This will give a density at most 1./(4*np.pi*(613.12/2)**3/3) = 8.286381584410062e-09

![hist-cpu-new](https://cloud.githubusercontent.com/assets/26856072/25990059/a185fcae-36fe-11e7-9495-15cf7c44715f.png)

Please let me know what else I can check. 
Thank you.

**Comment by weiguangcui** [comment]
Here is the config file
[config.h.zip](https://github.com/regonzar/paravt/files/995499/config.h.zip)

**Comment by regonzar** [comment]
Hi, i checked your data, and yes i made a mistake in the previous plot.
I ran the original version of the code, use the same config you have but borderfactor=3.0
I found you set BOX parameter but the header has "head problems_dens_sample_DM_sub2.txt"
4171 14000 14000 14000
you must change to
4171 14000
if you use box, 

i selected only central particles selecting a sphere with
xc=7552.
yc=6334.
zc=6693.
radius=3000.0
this result in 337 particles inside that radius

And i print the density of the first 5 particles

Here are my results:
2 cpu
dens = 0.00127800  2.00200e-05  1.75500e-06  9.20800e-07  8.33000e-07
pos = 4           0        2006        2007         920 

4cpu
0.00105100  2.00200e-05  1.75500e-06  9.20800e-07  8.33000e-07 
4           0        2006        2007         920 

8cpu
0.00105100  2.00200e-05  1.87500e-05  1.75500e-06  9.20800e-07
 4           0           1        2006        2007 

So yes, there are differencesm so i computed using borderfactor=10 and PERIODIC condition and i got:

2cpu
0.00127800  2.00200e-05  1.75500e-06  9.20800e-07  8.33000e-07 
4           0        2006        2007         920 
And EXACTLY the same for 4 and 8 cpu. so this is my conclusion:
1-fix the header of input data
2-your sample has too few particles so any split may lead to very long range neighbor connection, even for central particles in your distribution. so increasing borderfactor and seting periodic conditions to have constant conditions at borders fix the problem.

I recommend you to run on a larger sample where split data make sense.
I made several convergence test on large datasets and it works fine, but never tested on so few particles where in principle one particle may be neighbor of a particle in the other side of the volume.

Hope this help, i already spent some time on this, if you have more questions regarding you specific dataset send me an email to regonzar@astro.puc.cl
